src/services/codefixes/importFixes.ts

```diff
@@ -620,4 +666,73 @@ namespace ts.codefix {
         // Need `|| "_"` to ensure result isn't empty.
         return !isStringANonContextualKeyword(res) ? res || "_" : `_${res}`;
     }
+
+    function createLazyPackageJsonDependencyReader(fromFile: SourceFile, host: LanguageServiceHost) {
+        const packageJsonPaths = findPackageJsons(getDirectoryPath(fromFile.fileName), host);

```

Comment:
I think from tsserver perspective caching packageJson should be fairely ok since we do watch failed lookup locations and any changes in there result in recomputing program. But that's not guaranteed with other hosts since they can have their own module resolution or use the default resolution cache which doesn't watch these locations.

Reply:
> we do watch failed lookup locations and any changes in there result in recomputing program

Just to make sure I’ve interpreted this correctly: you’re saying that if a user runs `npm install new-package` which changes the package.json, `synchronizeHostData` will be called so I could update the cache there?

> But that's not guaranteed with other hosts

Maybe we can implement a fast check to determine whether the cache is up-to-date (last modified timestamp? content hash?). Actually, that might be better than reading eagerly on recomputing the program, since the program will change more often than the package.json.